Fix app redirection loop on browser's incognito mode and 3rd party cookie block

kimlisa opened the pull request:
resolves https://github.com/gravitational/teleport/issues/26009, https://github.com/gravitational/teleport/issues/15935

Updated RFD that explains the flow: https://github.com/gravitational/teleport/pull/31103

The cookie failure issue, is not a problem if an app is accessed by the default teleport subdomain. It becomes a problem if user sets a `public _addr` for the app where the domain is not teleports. 

Given SiteA (teleport) and SiteB (domain of app), the current implementation, responded with app cookies from SiteB, when the current site context was still on SiteA. To the browser this is considered 3rd party cookie and is blocked by default for the following major browsers (reproduced it too): 
- firefox: for both incognito and normal modes
- safari: for both incognito and normal modes
- brave
- chrome: only incognito, however, google will be [phasing out 3rd party cookie](https://developer.chrome.com/en/docs/privacy-sandbox/third-party-cookie-phase-out/) by mid next year (2024)

Our previous implementation that used the oauth state exchange flow was not affected by this problem because we set the app cookies while we are in the app domain.

This PR reverts some parts of the current implementation with the previous implementation with some small refactors, notably the state token race condition which was resolved by creating unique cookies for each state token created.

#### Tested
Tested the following by updating teleport versions (from my reproduction steps) that includes this PR's changes:

- with updated cluster and app service still running previous version
- github login

Local testing to test parameters are getting saved:
- redirecting with github login and local login with this URL: https://grafana-test-127-0-0-1.nip.io:3080/alerting/list?search=state:inactive%20type:alerting%20health:nodata

With Browsers:
- firefox both incognito and normal
- as suggested by [google](https://developer.chrome.com/en/docs/privacy-sandbox/third-party-cookie-phase-out/#debugging-cookies), tested in chrome by running the following in CLI `/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome --test-third-party-cookies-phaseout`

#### Reproduction
Running version: `Teleport Enterprise v15.0.0-dev git:pull-4093-g68d8647cd6 go1.21.3` (before this PR)

- Create two ec2 instances
- On the main cluster, install [certbot](https://certbot.eff.org/instructions?ws=webproduct&os=ubuntufocal), then create two certs (one for cluster and one for app) and include them in main teleport config under `proxy_service.https_keypairs` (note: need to add inbound rule for port 80)
- start teleport with config
- point custom domain (`kimlisa.app`) to this cluster's ip address (eg: `3.83.9.149`):
<img width="1335" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/323bf28b-c93b-4517-b148-edb508505198">


ec2 instance 1 (main cluster) cfg:
```
version: v3
teleport:
  nodename: ip-172-31-93-149
  data_dir: /var/lib/teleport
  log:
    output: stderr
    severity: INFO
    format:
      output: text
  ca_pin: ""
  diag_addr: ""
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  cluster_name: root.3.83.9.149.nip.io
  license_file: /var/lib/teleport/license.pem
  proxy_listener_mode: multiplex
  authentication:
    type: local
    second_factor: off
ssh_service:
  enabled: "yes"
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:443
  public_addr: root.3.83.9.149.nip.io:443
  https_keypairs:
    - key_file: /etc/letsencrypt/live/app1.kimlisa.app/privkey.pem
      cert_file: /etc/letsencrypt/live/app1.kimlisa.app/fullchain.pem
    - key_file: /etc/letsencrypt/live/root.3.83.9.149.nip.io/privkey.pem
      cert_file: /etc/letsencrypt/live/root.3.83.9.149.nip.io/fullchain.pem
  https_keypairs_reload_interval: 0s
  acme: {}
```

ec2 instance 2 for app.
- run whatever app (mine is running on `http://localhost:4000`)
- copy and paste the output from running `./tctl tokens add` in ec2 instance 1
- manually add a `public-addr` flag to custom domain

```
ubuntu@ip:~/teleport/e/build$ sudo ./tctl tokens add \
--type=app \
--app-name=test-app-1 \
--app-uri=http://localhost:4000
The invite token: 3c3...
This token will expire in 30 minutes.

Fill out and run this command on a node to make the application available:

> teleport app start \
   --token=3c3... \
   --ca-pin=sha256:80... \
   --auth-server=root.3.83.9.149.nip.io:443 \
   --name=test-app-1  \
   --uri=http://localhost:4000  \
   --public-addr=app1.kimlisa.app
```

Changelog: Fix app redirection loop on browser's incognito mode and 3rd party cookie block

kimlisa committed:
Copy pasta revert back to previous app redirection logic

Reverted parts of:
https://github.com/gravitational/teleport/pull/17592

Kept json field renames

kimlisa committed:
Remove unused fields (were renamed while back)

kimlisa committed:
Refactor previous implementation

- Create unique cookie for each state token created
- Preserve path and queries

kimlisa committed:
Split handlers

kimlisa committed:
Fix tests and lint

gzdunek commented:
Uh, this stuff is so complex. I looked through it and it looks fine, but I'm not very familiar that part, so I could have missed some things.

Also, do you think we need a security review here?

kimlisa commented:
> Also, do you think we need a security review here?

not likely, this PR is just a revert to the old implementation, the only thing that changed was that instead of overwriting our state token cookie with new requests, we create a unique state token cookie per request (short lived 1 minute)

kimlisa commented:
friendly ping @ryanclark

zmb3 commented on lib/web/app/auth.go:
```diff
@@ -18,45 +18,137 @@ package app
+
+	nonce, err := utils.CryptoRandomHex(auth.TokenLenBytes)
+	if err != nil {
+		h.log.WithError(err).Debugf("Failed to generate and encode random numbers.")
```
If I was running a Teleport cluster and I saw this in my logs I would have no idea what it means. Can we add a better message?

(Also, we use the exact same message above, so this adds addition burden for figuring out which line triggered the log message)

zmb3 commented on lib/web/app/auth.go:
```diff
@@ -18,45 +18,137 @@ package app
+	SetRedirectPageHeaders(w.Header(), nonce)
+
+	// Serving the HTML page.
+	fmt.Fprintf(w, appRedirectionJs, nonce)
```
nit: rename this constant

`appRedirectPage` or `appRedirectHTML` would work

zmb3 replied:
This is probably fine (we control the input), but you do have to be very careful using printf on HTML (it's prone to injection). Using `html/template` instead would be better, since it's HTML-aware and not a generic string replacement.

zmb3 commented on lib/web/app/auth.go:
```diff
@@ -18,45 +18,137 @@ package app
+
+// completeAppAuthExchange completes the auth exchange flow started by "startAppAuthExchange" handler
+// by validating the values passed in the request body, and upon success sets cookies required
+// for the current app session. User should be able to interact with the app now.
```
```suggestion
// for the current app session.
```

zmb3 commented on lib/web/app/auth.go:
```diff
@@ -18,45 +18,137 @@ package app
 	httplib.SetNoCacheHeaders(w.Header())
+	var req fragmentRequest
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		return trace.Wrap(err)
```
Since the primary reason this would fail is invalid JSON, what do you think about `trace.BadParameter` instead?

Or `trace.AccessDenied`, which is what you use below.

kimlisa commented:
Actually I will be updating the RFD to explain this flow to help make this whole flow more understandable

zmb3 commented on lib/web/app/auth.go:
```diff
@@ -18,45 +18,137 @@ package app
 
-	cookieValue := r.Header.Get("X-Cookie-Value")
-	if cookieValue == "" {
+	tokens := strings.Split(req.StateValue, "_")
+	if len(tokens) != 2 {
+		h.log.Warn("Request failed: request state token is not in the expected format")
 		return trace.AccessDenied("access denied")
 	}
+	secretToken := tokens[0]
+	cookieID := tokens[1]
```
```suggestion
	secretToken, cookieID, ok := strings.Cut(req.StateValue, "_")
	if !ok {
		h.log.Warn("Request failed: request state token is not in the expected format")
		return trace.AccessDenied("access denied")
	}
```

Optional refactor, up to you.

zmb3 commented on lib/web/app/handler.go:
```diff
@@ -128,8 +129,8 @@ func NewHandler(ctx context.Context, c *HandlerConfig) (*Handler, error) {
-	h.router.POST("/x-teleport-auth", makeRouterHandler(h.withCustomCORS(h.handleAuth)))
-	h.router.OPTIONS("/x-teleport-auth", makeRouterHandler(h.withCustomCORS(nil)))
+	h.router.GET("/x-teleport-auth", makeRouterHandler(h.startAppAuthExchange))
+	h.router.POST("/x-teleport-auth", makeRouterHandler(h.completeAppAuthExchange))
```
Do we need to keep the `OPTIONS` call available for a bit?

As written, app access may break during upgrades when not all proxies are on the same version.

kimlisa replied:
how long will proxy versions differ though? (when ryan's changes went it, the app breakage wasn't taken into account) 

if we were to handle preventing temporary breakage, I would need to make `POST /x-teleport-auth` differentiate between `cors request with custom headers` and `non-cors request with state query param and auth cookie`

kimlisa replied:
i added backwards compat starting from this commit: https://github.com/gravitational/teleport/pull/30321/commits/cc691d09f03b9261edd32854f96984cdda6d433b

tested with v15 build using new and legacy AppLauncher.tsx component with `https://grafana-test-127-0-0-1.nip.io:3080/alerting/list?search=state:inactive%20type:alerting%20health:nodata`

zmb3 commented on lib/web/app/handler.go:
```diff
@@ -541,61 +542,107 @@ const (
+	u := url.URL{
+		Scheme: "https",
+		Host:   proxyPublicAddr,
+		Path:   fmt.Sprintf("/web/launch/%s", hostname),
```
Use `url.JoinPath` instead of sprintf here?

phosphore replied:
Just wanted to point out that `url.JoinPath` performs URL path encoding and path normalization (i.e. resolution, removal of redundant slashes) while Sprintf takes everything as-is.
On a different note, what happens if the `path` parameter does not start with a `/`?

zmb3 commented on lib/web/app/handler.go:
```diff
@@ -541,61 +542,107 @@ const (
-		RawQuery: query,
+	// Presence of a stateToken means we are beginning an app auth exchange.
+	if req.stateToken != "" {
+		u.RawQuery = fmt.Sprintf("state=%s&path=%s", url.QueryEscape(req.stateToken), url.QueryEscape(req.path))
```
Similar comment - what you have is fine (you were careful to properly escape the values), but why not use `url.Values` to encode the string rather than sprintf?

```
var v url.Values
v.Add("state", req.stateToken)
v.Add("path", req.path)
u.RawQuery = v.Encode()
```

zmb3 commented on lib/web/app/handler.go:
```diff
@@ -541,61 +542,107 @@ const (
+			}
+		}
+
+		u.Path = path.Join(urlPath...)
```
`url.JoinPath`?

zmb3 commented on lib/web/app/handler.go:
```diff
@@ -541,61 +542,107 @@ const (
+		// into the query part of the URL.
+		query := fmt.Sprintf("path=%s", url.QueryEscape(r.URL.Path))
+		if len(r.URL.RawQuery) > 0 {
+			query = fmt.Sprintf("%s&query=%s", query, url.QueryEscape(r.URL.RawQuery))
```
Similar comments as above.

kimlisa committed:
Address CR

phosphore commented on lib/httplib/httpheaders.go:
```diff
@@ -210,31 +210,6 @@ func SetIndexContentSecurityPolicy(h http.Header, cfg proto.Features, urlPath st
 	h.Set("Content-Security-Policy", cspString)
```
Not be strictly related to this PR, but you should consider implementing a report-uri.

kimlisa committed:
Add backwards compatability

kimlisa committed:
Emit errs with invalid vals and attempt to delete session

kimlisa committed:
Update test

kimlisa commented on lib/web/app/auth.go:
```diff
@@ -18,45 +18,145 @@ package app
+	}
+	if subtle.ConstantTimeCompare([]byte(secretToken), []byte(stateCookie.Value)) != 1 {
+		h.log.Warn("Request failed: state token does not match.")
+		return trace.AccessDenied("access denied")
```
emitted errors and attempt deleting app session starting with this commit: https://github.com/gravitational/teleport/pull/30321/commits/d24ddbb353a44b6dafea55d0b86d68e3fe3f6b8f

ibeckermayer commented on lib/web/app/redirect.go:
```diff
@@ -60,3 +62,59 @@ func MetaRedirect(w http.ResponseWriter, redirectURL string) error {
 	return trace.Wrap(metaRedirectTemplate.Execute(w, redirectURL))
 }
+
+const appRedirectHTML = `
```
Consider something similar to the line above

https://github.com/gravitational/teleport/blob/de6bf60dce11e0051161d5044797a2d4354840cc/lib/web/app/redirect.go#L43

such that this template doesn't need to be re-parsed each time the respective endpoint is hit.

ibeckermayer reviewed (commented):
I would need to spend a lot of time understanding context to give this a useful review. I'm happy to do that if necessary, but it seems better suited to somebody more well acquainted with app access and/or your https://github.com/gravitational/teleport/pull/31103 RFD.

bl-nero commented on web/packages/teleport/src/AppLauncher/AppLauncher.tsx:
```diff
@@ -76,7 +54,33 @@ export function AppLauncher() {
+      const url = getXTeleportAuthUrl({ fqdn, port });
+      url.searchParams.set('state', stateToken);
+      url.searchParams.set('subject', session.subjectCookieValue);
+      url.hash = `#value=${session.cookieValue}`;
```
Are we sure that whatever there is in the `cookieValue` variable doesn't need to be URL-escaped?

kimlisa replied:
it should be safe since the `cookieValue` is the [session id](https://github.com/gravitational/teleport/blob/95aa537f681555f460bdee212f6c96a3a2c901c9/lib/auth/sessions.go#L100), which is a hex string

bl-nero commented:
@kimlisa  One general comment: I'm really impressed by this PR, it's really well-thought and VERY well commented. Though it's a very complex subject, the code wasn't at all difficult to understand. Just a handful of small issues, and you'll have my approval.

kimlisa commented on lib/web/app/auth.go:
```diff
@@ -111,3 +282,67 @@ func checkSubjectToken(subjectCookieValue string, ws types.WebSession) error {
+		Path:     "/",
+		HttpOnly: true,
+		Secure:   true,
+		SameSite: http.SameSiteNoneMode,
```
i think it is required because of the same reason stated here: https://github.com/gravitational/teleport/pull/30321/files#diff-379535a3878805eca5813cd9aa07db7a254cb885df354203105c4984eabafd83R181

kimlisa commented on lib/web/app/middleware.go:
```diff
@@ -65,11 +69,11 @@ func (h *Handler) withAuth(handler handlerAuthFunc) http.HandlerFunc {
 	// Cookies). Given this, we should only redirect to it when this format is
 	// already in use.
-	if !HasSession(r) {
+	if p.stateToken == "" && !HasSession(r) {
```
i think the error message, is missing a comma, does it make more sense if written like this:

```
"redirecting to launcher when using client certificate, is not valid"
```

you can access application through the CLI like this: https://goteleport.com/docs/application-access/cloud-apis/aws-console/#step-89-access-aws-cli

so i think from the CLI when you run into a forwaring [error](https://github.com/gravitational/teleport/blob/de6bf60dce11e0051161d5044797a2d4354840cc/lib/web/app/handler.go#L303), it wouldn't make sense to redirect the user to the launcher from the cli, so we just return an error instead 

but if the request came from the web, there'd be a cookie attached to it ([HasSession](https://github.com/gravitational/teleport/blob/95aa537f681555f460bdee212f6c96a3a2c901c9/lib/web/app/handler.go#L513)), and if there is a cookie, then we can redirect

i'll add a clarifying comment

kimlisa commented on lib/web/app/middleware.go:
```diff
@@ -171,3 +209,11 @@ type routerAuthFunc func(http.ResponseWriter, *http.Request, httprouter.Params,
+	clusterName string
+	publicAddr  string
+	stateToken  string
+	arn         string
```
added comments

bl-nero replied:
Oh, I didn't know we have a specialized logic for this. Thanks!

kimlisa commented on lib/web/app/handler.go:
```diff
@@ -293,7 +300,7 @@ func (h *Handler) handleForwardError(w http.ResponseWriter, req *http.Request, e
 	session, err := h.renewSession(req)
 	if err != nil {
-		if redirectErr := h.redirectToLauncher(w, req); redirectErr == nil {
+		if redirectErr := h.redirectToLauncher(w, req, launcherURLParams{}); redirectErr == nil {
```
the `redirectToLauncher` calls [makeAppRedirectUrl](https://github.com/gravitational/teleport/blob/de6bf60dce11e0051161d5044797a2d4354840cc/lib/web/app/handler.go#L659), which preserves the original path

kimlisa commented:
> @kimlisa One general comment: I'm really impressed by this PR, it's really well-thought and VERY well commented. Though it's a very complex subject, the code wasn't at all difficult to understand. Just a handful of small issues, and you'll have my approval.

@bl-nero don't be impressed. this PR is mostly a `revert` of a previous implementation (i think it was charn and alexey who were the original creators). i just renamed/restructured some things and added bunch of comments because everyone was confused (especially me) 😅

kimlisa committed:
Address CRs

kimlisa committed:
Fix lint

r0mant reviewed (approved):
@kimlisa Let's get this in so we can cover this during v15 test plan.

kimlisa committed:
Merge branch 'master' into lisa/revert-and-fix-app-redirection
